### PR TITLE
Json tweaks

### PR DESCRIPTION
**Pull request by thomasspriggs**
This is a collection of minor tweaks in `json.h`, to improve the usability of `json_array` and `json_string`.

### CODE REVIEW

File: src/util/json.h [hunk 1]
```diff
@@ -182,8 +184,18 @@ class json_arrayt:public jsont
...
 {
 public:
-  explicit json_stringt(const std::string &_value):
-    jsont(kindt::J_STRING, _value)
+  explicit json_stringt(std::string _value):
+    jsont(kindt::J_STRING, std::move(_value))
+  {
+  }
+
+  explicit json_stringt(const irep_idt &_value)
```
Comment (tautschnig) [inline src/util/json.h]: This requires `#ifndef USE_STD_STRING`.

**Review by tautschnig — CHANGES_REQUESTED** [review]
One fix is required as indicated by the message below (hence my request-for-changes), but what I'm unhappy about (though no blocker) are two aspects:
1) The commit messages repeatedly make use of "some" or "minor" without any proper quantification. These are weasel words. If you haven't take any measurements (which is really what should be done) then just omit those words.
2) Are there plans for any follow-up cleanup of code now that apparently the API is better? Otherwise: what's the benefit of this work?

**Comment by thomasspriggs** [comment]
These improvements stem from limitations I discovered whilst writing new code which produces json output. So the immediate benefit is that future json producing code can be better.
As a minimum I intend to revisit my new code to make use of these updates from cbmc. If this PR is merged in, I can add some clean up of other existing code to my todo list.
I used the word "minor", as I wanted to make it clear that I wasn't claiming substantial performance benefit. I will re-word the commit messages.

**Review by tautschnig — APPROVED** [review]

File: src/util/json.h [hunk 2]
```diff
@@ -169,13 +171,33 @@ class json_arrayt:public jsont
...
     return array.back();
   }
+
+  template <typename... argumentst>
+  void emplace_back(argumentst &&... arguments)
+  {
+    array.emplace_back(std::forward<argumentst>(arguments)...);
+  }
+
+  typedef jsont value_type; // NOLINT(readability/identifiers)
```
Comment (thk123) [inline src/util/json.h]: What's the point of this?

File: src/util/json.h [hunk 3]
```diff
@@ -169,13 +171,63 @@ class json_arrayt:public jsont
...
+  {
+    return array.end();
+  }
+
+  std::vector<jsont>::const_iterator cend() const
+  {
+    return array.cend();
+  }
+
+  typedef jsont value_type; // NOLINT(readability/identifiers)
```
Comment (thk123) [inline src/util/json.h:211]: What's the point of this?

**Review by thk123 — APPROVED** [review]

File: src/util/json.h (hunk 3 above)
Comment (thk123) [reply src/util/json.h:211]: To make more complaint with std algorithms

Comment (tautschnig) [reply src/util/json.h:211]: It might help to add some concrete examples in the commit message.

Comment (thomasspriggs) [reply src/util/json.h:211]: I discussed this with @thk123 in person. He hadn't actually read the existing message on this commit. I will however come up with a simple example to add to the commit message.

Comment (thomasspriggs) [reply src/util/json.h:211]: @tautschnig Added.

**Pull request merged**